[reland] use scatter in shard_tensor API

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75991

This is a reland of PR https://github.com/pytorch/pytorch/pull/72160. The previous PR failed on some cases where un-even scatter happens. So this PR made some additional fixes to ensure it scatters correctly.
1. Fix a bug in ProcessGroupNCCL::scatter, which is a similar issue to https://github.com/pytorch/pytorch/pull/75535
2. resize the shard to the same size before calling `dist.scatter`, and resize it back to the original layout after receiving from scatter.

Differential Revision: [D35726920](https://our.internmc.facebook.com/intern/diff/D35726920/)

Commit:
[reland] use scatter in shard_tensor API

This is a reland of PR https://github.com/pytorch/pytorch/pull/72160. The previous PR failed on some cases where un-even scatter happens. So this PR made some additional fixes to ensure it scatters correctly.
1. Fix a bug in ProcessGroupNCCL::scatter, which is a similar issue to https://github.com/pytorch/pytorch/pull/75535
2. resize the shard to the same size before calling `dist.scatter`, and resize it back to the original layout after receiving from scatter.

Differential Revision: [D35726920](https://our.internmc.facebook.com/intern/diff/D35726920/)

[ghstack-poisoned]

Commit:
Update on "[reland] use scatter in shard_tensor API"


This is a reland of PR https://github.com/pytorch/pytorch/pull/72160. The previous PR failed on some cases where un-even scatter happens. So this PR made some additional fixes to ensure it scatters correctly.
1. Fix a bug in ProcessGroupNCCL::scatter, which is a similar issue to https://github.com/pytorch/pytorch/pull/75535
2. resize the shard to the same size before calling `dist.scatter`, and resize it back to the original layout after receiving from scatter.

Differential Revision: [D35726920](https://our.internmc.facebook.com/intern/diff/D35726920/)

[ghstack-poisoned]

Commit:
Update on "[reland] use scatter in shard_tensor API"


This is a reland of PR https://github.com/pytorch/pytorch/pull/72160. The previous PR failed on some cases where un-even scatter happens. So this PR made some additional fixes to ensure it scatters correctly.
1. Fix a bug in ProcessGroupNCCL::scatter, which is a similar issue to https://github.com/pytorch/pytorch/pull/75535
2. resize the shard to the same size before calling `dist.scatter`, and resize it back to the original layout after receiving from scatter.

Differential Revision: [D35726920](https://our.internmc.facebook.com/intern/diff/D35726920/)

[ghstack-poisoned]

fduwjj commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
+
+        local_shards.append(Shard(tensor=local_tensor, metadata=local_metadata))
+
+        st = ShardedTensor._init_from_local_shards_and_global_metadata(
```
This will also fix the metadata order we have discussed about before, right? I mean when the placements are like: 
```
placements=[
    "rank:2/cuda:2",
    "rank:3/cuda:3",
    "rank:0/cuda:0",
    "rank:1/cuda:1",
]
```

wanchaol replied:
Yep it should fix it

fduwjj commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
+                    # resize the narrowed tensor to the same size and use it for
+                    # the scatter collective as dist.scatter requires same size
+                    # inputs on every rank
+                    tensor_to_scatter = narrowed_tensor.detach().clone().resize_(scatter_shape)
```
I tested it locally, this is essentially same as padding, random numbers are just added into the tensor for a bigger shape.. From the wiki of `resize_`, it says "Existing elements are preserved but any new memory is uninitialized.", I am not sure if there could be any side effect. If not, I think it's good to go.

wanchaol replied:
Yeah I also looked into the `gather_object`  and `all_gather_object` in the c10d API, seems like they are all using `resize_` ([link](https://github.com/pytorch/pytorch/blob/b369b89f235f54bc9de85d768fb62ac4579681dc/torch/distributed/distributed_c10d.py#L1761)) So I think we are good here.

I am even thinking maybe we should switch the `pad` and `unpad` to also use `resize_` to consolidate our approaches on the non-uniform inputs.

fduwjj replied:
I see. Is there difference in padding and resize? I mean which one is more efficient?

fduwjj reviewed (commented):
Overall, looks good to me but let's wait a little to see how others think of this solution. If no objection, I will stamp on it later this week. Also might be manually run some sharded op tests in case CI didn't capture potential failure. Thanks!

pritamdamania87 commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
+
+                tensors_to_scatter[rank] = tensor_to_scatter
+
+            if current_rank == rank:
```
nit: This is the same condition as above and can be merged?

wanchaol replied:
Hmm so it's a bit different, the first condition is `current_rank=src_rank` where it collects all the shards to scatter (only on src_rank), the second condition is `current_rank = rank` where it check if we are in the rank of the current `shard_metadata`.

pritamdamania87 commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -195,31 +193,69 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
+
+                tensors_to_scatter[rank] = tensor_to_scatter
+
+            if current_rank == rank:
```
nit: This is the same condition as above and can be merged?

Commit:
Update on "[reland] use scatter in shard_tensor API"


This is a reland of PR https://github.com/pytorch/pytorch/pull/72160. The previous PR failed on some cases where un-even scatter happens. So this PR made some additional fixes to ensure it scatters correctly.
1. Fix a bug in ProcessGroupNCCL::scatter, which is a similar issue to https://github.com/pytorch/pytorch/pull/75535
2. resize the shard to the same size before calling `dist.scatter`, and resize it back to the original layout after receiving from scatter.

Differential Revision: [D35726920](https://our.internmc.facebook.com/intern/diff/D35726920/)

[ghstack-poisoned]

Commit:
Update on "[reland] use scatter in shard_tensor API"


This is a reland of PR https://github.com/pytorch/pytorch/pull/72160. The previous PR failed on some cases where un-even scatter happens. So this PR made some additional fixes to ensure it scatters correctly.
1. Fix a bug in ProcessGroupNCCL::scatter, which is a similar issue to https://github.com/pytorch/pytorch/pull/75535
2. resize the shard to the same size before calling `dist.scatter`, and resize it back to the original layout after receiving from scatter.

Differential Revision: [D35726920](https://our.internmc.facebook.com/intern/diff/D35726920/)

[ghstack-poisoned]
